Fixed #31858 -- Reallowed whitespaces in URL paths outside of parameters.

timpark0807 opened the pull request:
**Ticket:** https://code.djangoproject.com/ticket/31858

**Background**
[Ticket 29667](https://github.com/django/django/pull/11688) implemented an exception if there was whitespace found in the URL path converter. 

This ticket updates that implementation by specifying invalid white spaces only between the `<` and `>` characters.  

**Implementation**
Instead of passing the entire `route` string into the exception check on Line 212, we only pass the part of the route between the`<` and `>` characters. To do this, we find the positions of those 2 characters and string index between them. 

`.find()` returns the index of the first occurrence of a character in a string.

![Screen Shot 2020-08-29 at 12 30 07 PM](https://user-images.githubusercontent.com/51100862/91644697-6d66cd80-e9f3-11ea-8dd7-a49780c6305e.png)


We could also do this in a 1 liner but that would sacrifice some readability. 
![Screen Shot 2020-08-29 at 12 32 58 PM](https://user-images.githubusercontent.com/51100862/91644738-cd5d7400-e9f3-11ea-9e62-1764ea1a92e0.png)


I also updated the error message to be more specific to this new implementation: 
Old Error: `URL route 'space/<int: num>' cannot contain whitespace.`
New Error: `URL route 'space/<int: num>' cannot contain whitespace between the < and > characters.`

smithdc1 commented:
I'm not sure this works where there are multiple parameters? 

A slight change to the test raises the error below. 
```python
    def test_space_in_route(self):
        msg = "URL route 'space/<int: num>' cannot contain whitespace between the < and > characters."
        with self.assertRaisesMessage(ImproperlyConfigured, msg):
            path('space/<int:num>/<int: num>', empty_view)
```

`space/<int:num>/<int:
num>' uses parameter name ' num' which isn't a valid Python identifier.`

timpark0807 commented:
> I'm not sure this works where there are multiple parameters?

@smithdc1 Thanks for the review! 

I'm seeing 2 ways we can handle multiple parameters within `route`. Let me know your thoughts! 

## **Option 1: String Splits**

Benefit of keeping the usage of `isdisjoint()`and fewer lines of code than option 2. The downside is it seems a bit hacky. 

**Code**
```
route = 'space/<int:num>/<int: num>/hello/<str: test>/world'

for chunk in route.split('<')[1:]:
    if not set(chunk.split('>')[0]).isdisjoint(string.whitespace):
        raise ImproperlyConfigured(
            "URL route '%s' cannot contain whitespace between the < and > characters."
            % route
        )
```
**Explanation** 
We split the `route` string on every `<` character
```
route = 'space/<int:num>/<int: num>/hello/<str: test>/world'
route.split('<') 
>>> ['space/', 'int:num>/', 'int: num>/hello/', 'str: test>/world']
```

Iterate through each chunk we split and split that chunk on the `>`.  The first element of that array is the text between the `<` and `>` characters. 
```
route[1:] 
>>> ['int:num>/', 'int: num>/hello/', 'str: test>/world']

chunk1 = 'int:num>/'
chunk1.split('>')
>>> ['int:num', '/']

chunk2 = 'int: num>/hello/'
chunk2.split('>')
>>> ['int: num', '/hello/']

chunk3 = 'int: num>'
chunk2.split('>')
>>> ['int: num', '']
```

## **Option 2: Stack**
Probably the more "formal" way to approach this problem. The downside is that it's a bit more confusing than option #1 if you haven't been solving Data Structures & Algorithm problems every single day for the last 8 months ;) 

```
route = 'space/<int:num>/<int:num>/hello/<str: test>/world'
stack = []

for char in route:
    if char != '>':
        stack.append(char)
    else:
        while stack and stack[-1] != '<':
            if stack.pop() == string.whitespace:
                raise ImproperlyConfigured(
                    "URL route '%s' cannot contain whitespace between the < and > characters."
                    % route
                )
```
**Explanation**
Iterate over the route string and push all characters that are not a closed `>`onto a stack (Last In First Out).

When we encounter a closed `>`, pop elements from the stack until we find it's matching `<` character. 

This means we can evaluate characters popped from the stack as the text between `<`and `>`. 

If we find a `string.whitespace` during this popping sequence, raise the Exception. 
If we encounter the matching `<` character, that means the text between `<`and `>` did not contain a whitespace. Continue on with iteration.

smithdc1 commented:
Thanks for your analysis. 👍 

Could we use `match` to do some of the work here, how about something like this? 🤔  What do you think? 

Note, was unsure why `string.whitespace` would not work in place of `' '`, but getting late here now. 

```Index: django/urls/resolvers.py
===================================================================
--- django/urls/resolvers.py	(revision e25d255260c3772a458a5bea63552ca186ab7704)
+++ django/urls/resolvers.py	(date 1598822469977)
@@ -219,6 +219,11 @@
     converters = {}
     while True:
         match = _PATH_PARAMETER_COMPONENT_RE.search(route)
+        if ' ' in match.group():
+            raise ImproperlyConfigured(
+                "URL route '%s' cannot contain whitespace between the < and > characters."
+                % original_route
+            )
         if not match:
             parts.append(re.escape(route))
             break
```

timpark0807 commented:
Awesome find! Taking advantage of `match` is definitely the cleanest approach here. 

I made a few minor edits to your solution in my newest commit. 

1. Running the test suite shows that `match` can resolve to `None` so let's check `match` exists first to prevent a `NoneType` error when calling `match.group()` 
2. Changed ImproperlyConfigured error return value from `% route` to `% original route` to handle case of multiple parameters e.g. `space/<int:num>/hello/<str: test>` (added a test for this as well)

```
django/urls/resolvers.py

def _route_to_regex(route, is_endpoint=False):
    original_route = route
    parts = ['^']
    converters = {}
    while True:
        match = _PATH_PARAMETER_COMPONENT_RE.search(route)
        if match and not set(match.group()).isdisjoint(string.whitespace):
            raise ImproperlyConfigured(
                "URL route '%s' cannot contain whitespace between the < and > characters."
                % original_route
        )
    ...
```

I went down a little rabbit hole and found the following definition for [string.whitespace](https://docs.python.org/3.1/library/string.html#string.whitespace) 
> A string containing all ASCII characters that are considered whitespace. This includes the characters space, tab, linefeed, return, formfeed, and vertical tab.

So string.whitespace is not _just_ `' '` but a combination of spaces, tabs, returns, etc. Now it makes sense why `string.whitespace` is not in `match.group()` and playing around in the Python shell helped preserve my sanity. 

Learned something new today!  

![Screen Shot 2020-08-30 at 3 22 48 PM](https://user-images.githubusercontent.com/51100862/91670701-b2aefc00-ead4-11ea-83c2-4391d19c2375.png)

Sticking with the original implementation of `set(route).isdisjoint(string.whitespace)` seems to be the safest and most encompassing solution since we are checking for more than just `' '`. 

Let me know if you see anything else. This looks good to me 👍🏼

smithdc1 commented:
Hi Tim,

Thanks for your efforts here, super clear explanations. The whitespace issue is obvious now!

I think this looks great. My only suggestion now is a couple of different tests.

- test that whitespace outside of the angle brackets is valid. 

- test for whitespace other than a `' '`. (Less sure about the need for this one)

timpark0807 commented:
> * test that whitespace outside of the angle brackets is valid.

@smithdc1 Good catch! I tested for this by checking that it returns a valid `URLPattern` type. Let me know if you see a cleaner way to handle this.

felixxm commented:
@timpark0807 Thanks :+1: I pushed minor edits and added test for other characters.

felixxm commented:
Merged in ece18207cbb64dd89014e279ac636a6c9829828e.
